interpreter/valid/match.ml

```diff
@@ -160,6 +160,19 @@ and match_deftype c dt1 dt2 =
     match_heaptype c (UseHT (Inexact, ut1)) (UseHT (Inexact, (Def dt2)))
   ) uts1
 
+let match_typeuse c ut1 ut2 (rt1 : rectype) (rt2 : rectype) =
```

Comment:
Passing in the rectypes looks like a hack to me. Any such side information should rather come from the context.

In the formal rules (e.g. in SpecTec), the context has another component `recs`, which bind recusive type indices. So far, this wasn't needed in the interpreter, as it was only needed to validate rolled deftypes, which only occurs in the soundness proof. Hence the interpreter omits it. But perhaps a version of this is what's needed here?

But I'm rather confused why that should be the case. Where are the rec type indices actually coming from that can enter this function? AFAICS, all calls either use unrolled or not yet rolled arguments, so I would expect no rec type index to occur. If you asserted false for them, would this fail?

Reply:
Indeed, it looks like everything still works out if the `Rec i` cases assert false. I guess I didn't consider that it would be ok for `match_typeuse` to be partial, and therefore had to pass `rt1` and `rt2` in to handle the recursive cases.